docs/tasks/administer-cluster/kubeadm-upgrade-ha.md

```diff
@@ -0,0 +1,179 @@
+
+Because on the worker nodes there is no keepalived to stop and start and all processes need to be moved to other nodes cleanly before shutting anything down, `drain` is used instead of `cordon`. In this process all containers will get moved from one node to another at least once. If some of them keep local data you will need some strategy for this scenario. This really depends on the application(s) running in the cluster, and there is no simple, generic solution to this.
+
+On the master (or wherever you run `kubectl` commands):
+
+```shell
+$ kubectl drain --delete-local-data --ignore-daemonsets <WORKER-NODE-FQDN>
+```
+
+On the respective worker node:
+
+```shell
+$ systemctl stop keepalived kubelet docker
+# use your distro's package manager, e.g. 'yum' on RH-based systems
+# for the versions stick to kubeadm's output (see above)
+$ yum install -y kubelet-<NEW-K8S-VERSION> kubectl-<NEW-K8S-VERSION> kubeadm-<NEW-K8S-VERSION> kubernetes-cni-<NEW-CNI-VERSION>
+$ systemctl start `echo keepalived kubelet docker | tac`
+```
+
+Like above, for _deb_-based systems use `apt-get install <PACKAGE>=version` to upgrade the packages to the respective desired versions.
+
+Now the new version of the `kubelet` should be running on the host. Verify this using the following command on the respective host:
+
+```shell
+systemctl status kubelet
+```
+
+If everything was successful, again on the master (or wherever you run `kubectl` commands):
+
+```shell
+$ kubectl uncordon <WORKER-NODE-FQDN>
+```
+
+After all hosts in your cluster have been upgraded, verify that all nodes are available again by executing the following from wherever you run `kubectl` commands:
+
+```shell
+kubectl get nodes
+```
+
+If the `STATUS` column of the above command shows `Ready` for all of your hosts, you are done.
+
+## If something goes wrong
+
+If the upgrade fails the situation afterwards depends on the phase in which things went wrong:
+
+1. If `/tmp/kubeadm upgrade apply` failed to upgrade the cluster it will try to perform a rollback. Hence if that happened on the first master, chances are pretty good that the cluster is still intact.
+
+   You can run `/tmp/kubeadm upgrade apply` again as it is idempotent and should eventually make sure the actual state is the desired state you are declaring. You can use `/tmp/kubeadm upgrade apply` to change a running cluster with `x.x.x --> x.x.x` with `--force`, which can be used to recover from a bad state.
+
+2. If `/tmp/kubeadm upgrade apply` on one of the secondary masters failed you still have a working, upgraded cluster, but with the secondary masters in a somewhat undefined condition. You will have to find out what went wrong and join the secondaries manually. As mentioned above, sometimes upgrading one of the secondary masters fails waiting for the restarted static pods first, but succeeds when the operation is simply repeated after a little pause of one or two minutes. 
```

Comment:
Is there any additional guidance we can offer the users here? Since we are externally hosting etcd, can we provide the user a way to re-install the master cleanly as an alternative?

Reply:
That looks like a good idea, however I have not tried this yet. Maybe we can make this a future PR?